curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -300,315 +301,345 @@ FileCacheManager::FindOrCreateChunkCacheManager(uint64_t index) {
+        LOG(WARNING) << "read inode = " << inodeWrapper->GetInodeId()
+                     << " retry = " << retry;
+        bthread_usleep(retryIntervalMs * retry * 1000);
+    } else {
+        LOG(WARNING) << "read inode = " << inodeWrapper->GetInodeId()
+                     << " retry = " << retry
+                     << ", reach max interval = " << maxIntervalMs << " ms";
+        bthread_usleep(maxIntervalMs * 1000);
     }
 
-    return readOffset;
+    return 0;
 }
 
-int FileCacheManager::ReadFromS3(const std::vector<S3ReadRequest> &requests,
-                                 std::vector<S3ReadResponse> *responses,
-                                 char* dataBuf, uint64_t fileLen) {
-    uint64_t chunkSize = s3ClientAdaptor_->GetChunkSize();
-    uint64_t blockSize = s3ClientAdaptor_->GetBlockSize();
-    std::vector<S3ReadRequest>::const_iterator iter = requests.begin();
-    std::atomic<uint64_t> pendingReq(0);
-    curve::common::CountDownEvent cond(1);
-    bool async = false;
-    // first is chunkIndex, second is vector chunkPos
-    std::map<uint64_t, std::vector<uint64_t>> dataCacheMap;
-    GetObjectAsyncCallBack cb =
-        [&](const S3Adapter *adapter,
-            const std::shared_ptr<GetObjectAsyncContext> &context) {
-            if (context->retCode == 0) {
-                pendingReq.fetch_sub(1, std::memory_order_relaxed);
-                cond.Signal();
-                return;
+int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
+                            char *dataBuf) {
+    // 1. read from memory cache
+    uint64_t actualReadLen = 0;
+    std::vector<ReadRequest> memCacheMissRequest;
+    ReadFromMemCache(offset, length, dataBuf, &actualReadLen,
+                     &memCacheMissRequest);
+    if (memCacheMissRequest.empty()) {
+        return actualReadLen;
+    }
+
+
+    // 2. read from s3 (localcache -> s3)
+    std::shared_ptr<InodeWrapper> inodeWrapper;
+    auto inodeManager = s3ClientAdaptor_->GetInodeCacheManager();
+    if (CURVEFS_ERROR::OK != inodeManager->GetInode(inodeId, inodeWrapper)) {
+        LOG(ERROR) << "get inode = " << inodeId << " fail";
+        return -1;
+    }
+
+    uint32_t retry = 0;
+    do {
+        // generate kv request
+        std::vector<S3ReadRequest> kvRequest;
+        GenerateKVReuqest(inodeWrapper, memCacheMissRequest, dataBuf,
+                          &kvRequest);
+
+        // read from s3
+        int ret = ReadKVRequest(kvRequest, dataBuf, inodeWrapper->GetLength());
+        if (ret >= 0) {
+            // read from s3 ok
+            break;
+        } else if (ret == -2) {
+            // TODO(@anybody): ret should replace the current number with a
+            // meaningful error code
+            //  read from s3 not exist
+            // 1. may be the metaserver compaction update inode is not
+            //    synchronized to the client. clear inodecache && get agin
```

Comment:
get again

Reply:
Yes. ReadKVRequest is already handled with while.
The situations of s3 return error will retry：If the error is not exist, it may be that the data fragment has not been transmitted to s3, and it will wait for retry according to `HandleReadS3NotExist`. Other errors indicate that reading from s3 failed and will not be retried.